### PR TITLE
[tests-only][full-ci] add test to upload file to a shared folder via TUS by federated user

### PR DESCRIPTION
**Pull request by prashant-gurung899**
## Description
This PR adds test to upload a file to a shared folder via TUS by a federated user.
``` feature
Scenario: federated user upload file to a shared folder via TUS
```

## Related Issue

- Part of: https://github.com/owncloud/ocis/issues/10073
- Bug Report: https://github.com/owncloud/ocis/issues/10285

## How Has This Been Tested?

- locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/acceptance/features/apiOcm/share.feature [hunk 1]
```diff
@@ -618,3 +618,33 @@ Feature: an user shares resources using ScienceMesh application
...
+      | resource        | FOLDER                        |
+      | space           | Personal                      |
+      | sharee          | Brian                         |
+      | shareType       | user                          |
+      | permissionsRole | Editor                        |
+      | federatedServer | @federation-ocis-server:10200 |
+    And using server "REMOTE"
+#    When user "Brian" lists the shares shared with him using the Graph API
+#    Then the HTTP status code should be "200"
+    And user "Brian" has created a new TUS resource in the space "Shares" with the following headers:
```
Comment (saw-jan) [inline tests/acceptance/features/apiOcm/share.feature:640]: looks like the federated share is not accessible similar to normal (local) shares. so we might need to use different step for accessible federation shares.

- Cannot access federation shares via webdav paths. (GET, PUT, PROPFIND) doesn't work

Comment (saw-jan) [reply tests/acceptance/features/apiOcm/share.feature:640]: See: https://github.com/owncloud/ocis/issues/10536

Comment (rhafer) [reply tests/acceptance/features/apiOcm/share.feature:640]: Can you please add examples what hat requests exactly are failing?

Comment (saw-jan) [inline tests/acceptance/features/apiOcm/share.feature:640]: Could you try a step something like this?
```feature
When user "Alice" uploads a file with content "lorem" to "textFile.txt" inside federated share "FOLDER" via TUS using the WebDAV API
```
The step should get the `remoteItem.id` by requesting `sharedWithMe`, and use the id to do the upload request
`remote.php/dav/spaces/{remote-item-id}/textFile.txt`
```json
{
    "createdBy": {
        "user": {
            "@libre.graph.userType": "Federated",
            "displayName": "Admin",
        }
    },
    "name": "Parent",
    "remoteItem": {
        "id": "89f37a33-858b-45fa-8890-a1f2b27d90e1$aa43c7d3-62cc-4472-b2ce-e7b14b6cff9d!Lw==",
        "name": "Parent",
    }
}
```

Comment (prashant-gurung899) [reply tests/acceptance/features/apiOcm/share.feature:640]: @saw-jan 
Is it similiar to what the following existing step does? If yes then we can use it:
``` feature
When user "Brian" uploads a file with content "uploaded content" to "/FOLDER/file.txt" via TUS inside of the space "Shares" using the WebDAV API
```

Comment (saw-jan) [reply tests/acceptance/features/apiOcm/share.feature:640]: No, the endpoint to request will be a bit different. I referenced the new step from this one.

File: tests/acceptance/bootstrap/SpacesTUSContext.php [hunk 2]
```diff
@@ -148,19 +147,36 @@ public function userCreatesANewTusResourceForTheSpaceUsingTheWebdavApiWithTheseH
...
-			$this->featureContext->setLastUploadDeleteTime(\time());
-		} catch (Exception $e) {
-			Assert::assertStringContainsString('Unable to create resource', (string)$e);
+
+			$jsonBody = $this->featureContext->getJsonDecodedResponseBodyContent($response);
+
+			// trimming the resource 'FOLDER/file.txt' to get the destination path 'file.txt'
+			// as $spaceId below represents the remoteItem id
+			$resource = \ltrim($resource, "/");
+			$resource = preg_replace("#^[^/]+/#", "", $resource);
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesTUSContext.php:167]:
```suggestion
			$resources = \explode("/", $resouce);
			if (\count($resource) > 1) {
			  \array_shift($resources);
			  $resource = \join("/", $resources);
			}
```

File: tests/acceptance/bootstrap/SpacesTUSContext.php [hunk 3]
```diff
@@ -148,19 +147,36 @@ public function userCreatesANewTusResourceForTheSpaceUsingTheWebdavApiWithTheseH
...
+				$credentials['password']
 			);
-			$this->featureContext->setLastUploadDeleteTime(\time());
-		} catch (Exception $e) {
-			Assert::assertStringContainsString('Unable to create resource', (string)$e);
+
+			$jsonBody = $this->featureContext->getJsonDecodedResponseBodyContent($response);
+
+			// trimming the resource 'FOLDER/file.txt' to get the destination path 'file.txt'
+			// as $spaceId below represents the remoteItem id
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesTUSContext.php:165]:
```suggestion
			// trimming the resource 'FOLDER/file.txt' to get the destination path 'file.txt'
			// 'FOLDER' is represented by remoteItem id in the federation share
```

File: tests/acceptance/bootstrap/SpacesTUSContext.php [hunk 4]
```diff
@@ -148,19 +147,36 @@ public function userCreatesANewTusResourceForTheSpaceUsingTheWebdavApiWithTheseH
...
+			$resource = \ltrim($resource, "/");
+			$resource = preg_replace("#^[^/]+/#", "", $resource);
+			$spaceId = $jsonBody->value[0]->remoteItem->id;
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesTUSContext.php:168]:
```suggestion
			$share = <find share by name from $jsonBody->value>
			$spaceId = $share->remoteItem->id;
```

File: tests/acceptance/features/apiOcm/share.feature [hunk 5]
```diff
@@ -611,3 +611,21 @@ Feature: an user shares resources using ScienceMesh application
         }
       }
       """
+
+  @issue-10285
```
Comment (saw-jan) [inline tests/acceptance/features/apiOcm/share.feature:615]:
```suggestion
  @issue-10285 @issue-10536
```

File: tests/acceptance/bootstrap/SpacesTUSContext.php [hunk 6]
```diff
@@ -148,19 +147,50 @@ public function userCreatesANewTusResourceForTheSpaceUsingTheWebdavApiWithTheseH
...
-		try {
-			$this->tusContext->uploadFileUsingTus(
-				$user,
-				\basename($tmpFile),
-				$resource,
-				$spaceId
+		$server = $this->featureContext->getCurrentServer();
+
+		//check if it is REMOTE server and Shares space
+		if ($server === 'REMOTE' && $spaceName === 'Shares') {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesTUSContext.php:153]:
```suggestion
		if ($server === 'REMOTE' && $spaceName === 'Shares') {
```

Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesTUSContext.php:153]: $server is used only in the if statement
```suggestion
		if ($this->featureContext->getCurrentServer() === 'REMOTE' && $spaceName === 'Shares') {
```

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 7]
```diff
@@ -254,22 +254,67 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
...
+
+		$jsonBody = $this->featureContext->getJsonDecodedResponseBodyContent($response);
+
+		// Search for the folder name in the response to find the remoteItem ID
+		foreach ($jsonBody->value as $item) {
+			if (isset($item->name) && $item->name === $share) {
+				if (isset($item->remoteItem->id)) {
+					return $item->remoteItem->id;
+				} else {
+					throw new Exception("Failed to find remoteItem ID for share: $share.");
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesContext.php:284]:
```suggestion
					throw new Exception("Failed to find remoteItem ID for share: $share");
```

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 8]
```diff
@@ -254,22 +254,67 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
...
 	 * @param string $spaceName
 	 * @param string $fileName
+	 * @param string|null $share
 	 *
 	 * @return ResponseInterface
 	 * @throws GuzzleException
 	 */
-	public function getFileData(string $user, string $spaceName, string $fileName): ResponseInterface {
-		$space = $this->getSpaceByName($user, $spaceName);
+	public function getFileData(string $user, string $spaceName, string $fileName, ?string $share= null): ResponseInterface {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesContext.php:305]:
```suggestion
	public function getFileData(string $user, string $spaceName, string $fileName, ?bool $federatedShare = false): ResponseInterface {
```

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 9]
```diff
@@ -254,22 +254,67 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
...
 		$baseUrl = $this->featureContext->getBaseUrl();
-		$davPath = WebdavHelper::getDavPath(WebDavHelper::DAV_VERSION_SPACES, $space["id"]);
-		$fullUrl = "$baseUrl/$davPath/$fileName";
+
+		if ($this->featureContext->getCurrentServer() === "REMOTE" && $spaceName === 'Shares') {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesContext.php:308]:
```suggestion
		if ($federatedShare) {
```

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 10]
```diff
@@ -254,22 +254,67 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
...
+			$remoteItemId = $this->getSharesRemoteItemId($user, $share);
+			$remoteItemId = \rawurlencode($remoteItemId);
+			$davPath = WebdavHelper::getDavPath(WebDavHelper::DAV_VERSION_SPACES, $remoteItemId);
+			$fullUrl = "$baseUrl/$davPath/$fileName";
+		} else {
+			$space = $this->getSpaceByName($user, $spaceName);
+			$davPath = WebdavHelper::getDavPath(WebDavHelper::DAV_VERSION_SPACES, $space["id"]);
+			$fullUrl = "$baseUrl/$davPath/$fileName";
+		}
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesContext.php:318]:
```suggestion
			$spaceId = \rawurlencode($remoteItemId);
		} else {
			$space = $this->getSpaceByName($user, $spaceName);
			$spaceId = $space["id"];
		}
		$davPath = WebdavHelper::getDavPath(WebDavHelper::DAV_VERSION_SPACES, $spaceId);
		$fullUrl = "$baseUrl/$davPath/$fileName";

```

File: tests/acceptance/features/apiOcm/share.feature [hunk 11]
```diff
@@ -611,3 +611,22 @@ Feature: an user shares resources using ScienceMesh application
...
+    And user "Alice" has created a folder "FOLDER" in space "Personal"
+    And user "Alice" has sent the following resource share invitation to federated user:
+      | resource        | FOLDER                        |
+      | space           | Personal                      |
+      | sharee          | Brian                         |
+      | shareType       | user                          |
+      | permissionsRole | Editor                        |
+    When using server "REMOTE"
+    And user "Brian" uploads a file with content "lorem" to "file.txt" inside federated share "FOLDER" via TUS using the WebDAV API
+    Then for user "Brian" the content of the federated share file "FOLDER/file.txt" of space "Shares" should be "lorem"
```
Comment (saw-jan) [inline tests/acceptance/features/apiOcm/share.feature:632]:
```suggestion
    Then for user "Brian" the content of file "file.txt" of federated share "Folder" should be "lorem"
```

File: tests/acceptance/bootstrap/SpacesContext.php (hunk 8 above)
Comment (prashant-gurung899) [reply tests/acceptance/bootstrap/SpacesContext.php]: @saw-jan
Why are we setting `$federatedShare` as as boolean?

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 12]
```diff
@@ -254,21 +254,65 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
...
+		$jsonBody = $this->featureContext->getJsonDecodedResponseBodyContent($response);
+
+		// Search for the folder name in the response to find the remoteItem ID
+		foreach ($jsonBody->value as $item) {
+			if (isset($item->name) && $item->name === $share) {
+				if (isset($item->remoteItem->id)) {
+					return $item->remoteItem->id;
+				} else {
+					throw new Exception("Failed to find remoteItem ID for share: $share");
+				}
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesContext.php:285]:
```suggestion
				}
				throw new Exception("Failed to find remoteItem ID for share: $share");
```

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 13]
```diff
@@ -254,21 +254,65 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
 		return $space["id"];
 	}
 
+	/**
+	 * Retrieves the remoteItem ID for a given share in the Shares space.
+	 *
+	 * @param string $user
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/SpacesContext.php:260]: method name is clear
```suggestion
	 * @param string $user
```

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 14]
```diff
@@ -874,6 +918,32 @@ public function checkFileContent(
...
+		string $file,
+		string $share,
+		string $fileContent
+	): void {
+		// trimming the resource 'FOLDER/file.txt' to get the destination path 'file.txt'
+		// 'FOLDER' is represented by remoteItem id in the federation
+		$file = \ltrim($file, "/");
+
+		$actualFileContent = $this->getFileData($user, $share, $file, true)->getBody()->getContents();
+		Assert::assertEquals($fileContent, $actualFileContent, "$file does not contain $fileContent");
```
Comment (amrita-shrestha) [inline tests/acceptance/bootstrap/SpacesContext.php:944]: I think this way, when a test fails, it is easy to get an idea about failure. In past, we faced similar problem  regarding insufficient detail on error message 
```suggestion
		Assert::assertEquals($fileContent, $actualFileContent, "Expected file to content to be $fileContent but found $actualFileContent");
```

File: tests/acceptance/features/apiOcm/share.feature [hunk 15]
```diff
@@ -695,4 +695,23 @@ Feature: an user shares resources using ScienceMesh application
           }
         }
       }
-      """
\ No newline at end of file
+      """
+
+  @issue-10285 @issue-10536
+  Scenario: local user upload file to a federated shared folder via TUS
+    Given using spaces DAV path
```
Comment (amrita-shrestha) [inline tests/acceptance/features/apiOcm/share.feature:702]: does this matter about spaces dav path in this scenario?

File: tests/acceptance/features/apiOcm/share.feature [hunk 16]
```diff
@@ -695,4 +695,23 @@ Feature: an user shares resources using ScienceMesh application
           }
         }
       }
-      """
\ No newline at end of file
+      """
+
+  @issue-10285 @issue-10536
+  Scenario: local user upload file to a federated shared folder via TUS
```
[the remark this answers is not in the document]
Comment (prashant-gurung899) [reply tests/acceptance/features/apiOcm/share.feature:701]: If that's the case, shall we write two test scenarios?

File: tests/acceptance/features/apiOcm/share.feature (hunk 15 above)
Comment (prashant-gurung899) [reply tests/acceptance/features/apiOcm/share.feature:702]: Yes it matters, as we need to hit endpoint with spaceId.
:https://github.com/owncloud/ocis/blob/9b8b92c72fdbbb81951defdffa896d22b05e33cb/tests/acceptance/bootstrap/TUSContext.php#L310

Comment (saw-jan) [reply tests/acceptance/features/apiOcm/share.feature:702]: we must specify space dav path for federation shares (doesn't work with other dav paths, see https://github.com/owncloud/ocis/issues/10536). by default new dav path is used in the tests

File: tests/acceptance/bootstrap/SpacesContext.php [hunk 17]
```diff
@@ -254,21 +254,66 @@ public function getSpaceIdByName(string $user, string $spaceName): string {
...
+	 * Retrieves the remoteItem ID for a given share in the Shares space.
+	 *
+	 * @param string $user
+	 * @param string $share //share
+	 *
+	 * @return string
+	 *
+	 * @throws Exception|GuzzleException
+	 */
+	public function getSharesRemoteItemId(string $user, string $share): string {
```
[the remark this answers is not in the document]
Comment (saw-jan) [reply tests/acceptance/bootstrap/SpacesContext.php:265]: hmm, `$item->remoteItem->id` should always be a string. if not found the function will throw

**Review by saw-jan — APPROVED** [review]
it's :+1: for me.

**Pull request merged**